### DISCUSSION TITLE
Frontend/Backend Separation

### DISCUSSION THREAD

**Discussion by mbway** [mozman/ezdxf#534] [Ideas]
There seems to be some leakage between the responsibilities of the drawing frontend and the drawing backend. Ideally the backend should not have to understand anything about what is being drawn as it's the frontend's job to interpret the DXF entities and break them down to primitives that can be drawn by any backend. By enforcing this separation, it makes creating new backends easier because there is no hidden information; everything it needs is provided by the frontend.

the `DEFAULT_PARAMS` and attributes of the backend seem to violate this separation because some parameters directly relate to the dxf entities (eg pdsize, pdmode, show_defpoints, show_hatch, hatch_pattern). The backend should not have to understand what a defpoint or a hatch is. Any auxiliary information such as the fill pattern for a polygon should be passed as a parameter from the frontend.
Some parameters like the line weight makes sense because that is a parameter of the rendering unrelated to the CAD. It's the frontend's job to decide that a line should be drawn and if the CAD does not specify the line weight then it's up to the backend to decide how to display that line.

It seems like some backend attributes are being used as storage for the frontend to write into. Is there any reason why these attributes weren't added to the frontend directly?

**Comment by mozman** [comment]
Yes this is a mixed usage with the idea of a central configuration structure for the drawing add-on.

You can move some of the parameters to the frontend and the library user has to pass the custom params to the frontend and to the backend:
```Python
out = MatplotlibBackend(..., params=MYPARAMS)
frontend = Frontend(out, params=MYPARAMS)
```

If you don't plan to implement an additional backend, the only who will ever be affected by this issue is myself in the distant future (and I may hate myself for this). Maybe I will refactor this by myself **if** I add an SVG backend after the 1.0 release. 

pdsize, pdmode: handled by frontend
A point size (pdsize) < 0 means relative to viewport size and only the backend knows the current window size (not supported).

show_defpoints toggle: handled by frontend

show_hatch toggle: handled by frontend

hatch_pattern: handled by backend
The frontend has no knowledge of the capabilities of the backend to handle hatch pattern and this option defines what the backend should do if it can't render hatch pattern.

A better solution would be to write a hatch pattern renderer which breaks pattern down into lines, but I don't do it :). 
If someone else does it this should be a Cython extension, a pure Python implementation is maybe be too slow.

linetype_renderer, linetype_scaling, lineweight_scaling, min_line_weight, min_dash_length: handled by backend
Parameters for the backend specific line type renderer.

**Reply by mbway** [reply]
so is the current implementation aiming for convenience by only having to pass parameters to the backend to configure both the backend and the frontend?

> If you don't plan to implement an additional backend

I have a 'backend' which I currently maintain for dxf format conversion to a custom data format, although I'm ignoring most of these parameters as they are irrelevant for me. The current approach isn't necessarily a problem, just sub-optimal as it makes the backend more complex than necessary. It should be a black box from the point of view of the frontend, not an extension of it.

I think something like using a configuration dataclass (which would be possible since you already require python 3.7+) passed to the frontend by the user could be a good way to go because it gives strong typing rather than a dictionary and still only has to be passed in in one place. Then a method like `Backend.configure(params)` could be called to pass the parameters (either all of them or just the ones relevant to the backend) from the frontend to the backend before it starts drawing, then the backend can choose what it wants to do with the parameters (i.e. store and use them or just ignore them) and the frontend doesn't have to care either way. Then the backend should have no public attributes so it's clear that data only flows in one direction (front to back) and only via method calls rather than direct attribute access.

If I find the time I might have a go at refactoring this as it shouldn't be too much work
